Adding support to Indian Numbering system

praveentiru opened the pull request:
This is draft PR to fix #36. Changes in the code are captured below:

1. Added NumericSystem type to capture locale numeric type
2. Added numericSystem field to Locale to track the value
3. Added default numericSystem value to base Locale to preserve backward
   compatibility
4. Added functions to Parser to support splitting by Indian system
5. Updated parser function to support Indian numbering system

@cuducos I am assuming that most users of library use one of the locales such as base, usLocale etc. to build their locale rather than defining it from scratch. I am yet to update fromString function to get locale from browser. This PR is to enable you to review and advise me on my approach.

praveentiru committed:
Adding support to Indian Numbering system:

1. Added NumericSystem type to capture locale numeric type
2. Added numericSystem field to Locale to track the value
3. Added default numericSystem value to base Locale to preserve backward
   compatibility
4. Added functions to Parser to support splitting by Indian system
5. Updated parser function to support Indian numbering system

praveentiru commented:
FYI, I was not able to run tests as they are set-up. So, I manually tested the test cases I have added in comments of parse function. Trying to figure out how to run tests.

praveentiru commented:
@cuducos I was able to run tests. Lot of tests of format function are failing because of change in Locale data type. My assumption was locales would be built using one of the locales provided.

If that is not the case then, I have to figure alternate solution for same.

cuducos commented on src/FormatNumber/Locales.elm:
```diff
@@ -44,6 +45,18 @@ type Decimals
+{-| The 'NumericSystem' type contains different numeric systems currently 
+supported:
+
+  - Western - It separates digits by 1000s. 1000000 -> 1,000,000
+  - Indian - It separates digits by 1000 then it is separated by 100s. 
+    1000000 -> 10,00,000
```
For consistency we could use the same formatting and style as in `Decimals`:

```suggestion
  - `Western` separates digits by thousands (e.g.: 1000000 might be separated
    as 1,000,000).
  - `Indian` separates digits by the thousand and, from thaere, bu hundreds
    (e.g.: 1000000 might be separated as 10,00,000).

```

praveentiru replied:
I will the change in next commit from my fork.

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -106,6 +106,41 @@ splitThousands integers =
+    splitByIndian "12" --> [ "12" ]
+
+-}
+splitByIndian : String -> List String
+splitByIndian integers =
```
For consistency, should we rename `splitThousands` as `splitByWestern`?

praveentiru replied:
Yes, it can be template if we have to add support for more numbering systems.

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -106,6 +106,41 @@ splitThousands integers =
+            else
+                integers
+    in
+    List.reverse
+        <| thousand :: (integers |> String.dropRight 3 |> reversedSplitHundreds)
```
What about enhancing readability by using only one direction of `|>` or `<|` in the same expression?

Not sure if this is accurate, but I was wondeirng something like:

```suggestion
    integers
        |> String.dropRight 3
        |> reversedSplitHundreds
        |> (::) thousand
        |> List.reverse
```

praveentiru replied:
Agreed. I wanted to build it similarly but could not come up with this sequence. :)

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -106,6 +106,41 @@ splitThousands integers =
         |> reversedSplitThousands
         |> List.reverse
 
+{-| Split a `String` in `List String` grouping by digits as per Indian 
+numbering system. Last 3 digits are grouped together but after that 
+numbers are grouped in two. 
+[Indian numbering system](https://en.wikipedia.org/wiki/Indian_numbering_system#Use_of_separators):
+
+    splitByIndian "12345678" --> [ "1", "23", "45", "678" ]
+
+    splitByIndian "12" --> [ "12" ]
+
+-}
```
Formatted for consistency in style — please avoid leading whitespaces too ; )

```suggestion
{-| Splits `String` into `List String` grouping by digits as `Indian`. Last 3
digits are grouped together but after that numbers are grouped in pairs. For
more details:
[Indian numbering system](https://en.wikipedia.org/wiki/Indian_numbering_system#Use_of_separators).

    splitByIndian "12345678" --> [ "1", "23", "45", "678" ]

    splitByIndian "12" --> [ "12" ]


-}
```

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -210,6 +245,26 @@ getDecimals locale digits =
         Min min ->
             addZerosToFit min digits
 
+{-| Given a 'NumericSystem` parses a integer `String` into 
+a `List String` representing grouped integers:
+
+    - Western NumericSystem: 1000000 -> 1,000,000
+    - Indian NumericSystem: 1000000 -> 10,00,000
+-}
+
```
This is very problematic. Some points to consider:

* There is a typo (a ' instead of a ` NumericSystem, first line of the block)
* The output examples do not look like `List String`
* The input examples do not look like `String`
* There are no tests/examples that run with `elm-verify-tests`
* There are leading whitespace characters at the end of al ine
* There is an extra line between the docs/comment block and the declaration of the function

What about something among these lines?

```suggestion
{-| Given a `NumericSystem` parses a `String` into a `List String` representing
grouped integers:

    splitIntegers Western "1000000" --> ["1", "000", "000"]

    splitIntegers Indian "1000000" -> ["10", "00", "000']

-}
```

praveentiru replied:
I will make these changes from my fork as the second test is missing a hyphen.

cuducos reviewed (changes requested):
Great job! You're quick and your code is good : ) Congrats!

I made minor comments, mostly regarding the docs writing style and naming. What do you think about them?

Regarding your question about the tests, I am not 100% sure because you have not shared any error/failure message (and I have had no time to actually run your branch), however… if you change the `Locale` data type (and you did it by adding the `NumericSystem` to it) you have to update the `Locale` in every test that has a `Locale` as an input. Also, you might double-check if you update the `Locale` definitions (`usLocale`, `spacnishLocale` etc.) to make sure they have a `NumericSystem`). That's a little bit of the price we pay for the safety Elm and static type give us : )

praveentiru commented:
@cuducos Thanks for such quick review. I will work through your comments tomorrow my time. But, I have a big question. Are you ok with the change proposed?

The proposed change can break certain code. Please find below two identical cases
1. `Locale (Exact 2) "," "." "(" ")" "" "" "" ""` 
2. `{ base | decimals = Exact 2, thousandSeparator = ",", negativePrefix = "(", negativeSuffix = ")"}` 

In case 1, my change will break the code and in case 2 it will not as I have added `numericSystem` to `base`.

As deduced by you, if I replace the `Locale` built using base, or, usLocale they will go away.

cuducos commented:
> Are you ok with the change proposed?

Sure thing. I think the `NumericSystem` addition to `Locale` expands the possibilities of the package in a good way. Yes, this is a breaking change, but thanks we have semantic versioning… so I am happy to continue on this pathway.

Since we change the `Locale` type, we need to:

1. Make sure our suggested locales include the new `NumericSystem` field; probably adding `Western` to [`base`](https://github.com/cuducos/elm-format-number/blob/master/src/FormatNumber/Locales.elm#L76-L86) will do it, because the other locales inherit from it
1. In test cases, whenever we define a `Locale` we need to pass a `NumericSystem` to it

Example (on the 2nd item), where we have `Locale (Exact 2) "," "." "(" ")" "" "" "" ""` we would end up with `Locale (Exact 2) Western "," "." "(" ")" "" "" "" ""` (depending on where we add `NumericSystem` to the `Locale` record). Tests that use locales that are build from the `base` locale should not need any adjustment.

Does that make sense?

praveentiru committed:
Apply suggestions from code review

I have added two of your suggestions as-is. Rest I will be modify in my fork and commit them.

Co-authored-by: Eduardo Cuducos <4732915+cuducos@users.noreply.github.com>

praveentiru committed:
Changes in the commit:

Locales.elm
1. Updated comments for NumericSystem for consistency with Decimals
Parser.elm
1. Exposed additional functions in module declaration to run tests
   properly
2. Changed splitThousands to splitByWestern for consistency
3. Changed comments of splitByIndian for consistency
4. Updated piping in splitByIndian for better readability
5. Added tests for splitIntegers in comments
6. Fix tests for parse function by adding import statement
FormatNumber.elm
1. Fix tests by adding numericSystem where it is needed
Stringfy.elm
1. Fix tests by adding numericSystem where it is needed

praveentiru committed:
Merge branch 'master' of github.com:praveentiru/elm-format-number

praveentiru commented:
@cuducos I have made changes as per your comments. Please review. I still have to update fromString function to support Indian numbering system.

cuducos commented on src/FormatNumber/Locales.elm:
```diff
@@ -44,6 +45,19 @@ type Decimals
     | Max Int
     | Exact Int
 
+{-| The 'NumericSystem' type contains different numeric systems currently 
```
```suggestion
{-| The `NumericSystem` type contains different numeric systems currently
```

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -106,6 +108,45 @@ splitThousands integers =
+    splitByIndian "12345678" --> [ "1", "23", "45", "678" ]
+
+    splitByIndian "12" --> [ "12" ]
+
+
+-}
```
```suggestion

-}
```

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -210,10 +251,34 @@ getDecimals locale digits =
         Min min ->
             addZerosToFit min digits
 
+{-| Given a 'NumericSystem` parses a integer `String` into 
```
```suggestion
{-| Given a 'NumericSystem` parses an integer `String` into
```

cuducos commented on src/FormatNumber/Parser.elm:
```diff
@@ -210,10 +251,34 @@ getDecimals locale digits =
+
+    splitIntegers Indian "1000000" --> ["10", "00", "000"]
+
+-}
+
+splitIntegers : NumericSystem -> String -> List String
```
```suggestion
-}
splitIntegers : NumericSystem -> String -> List String
```

cuducos reviewed (commented):
Looking great 💜 Good job! Thank you so much.

I'll wait for `fromString` to merge, and then I can prepare a release. Feel free to rebase/squash your commits into a single one —or I merge using squash, if you don't mind.

praveentiru committed:
Apply suggestions from code review

Accepting all suggestions proposed

Co-authored-by: Eduardo Cuducos <4732915+cuducos@users.noreply.github.com>

cuducos committed:
Merge branch 'indian-numeric-system' into master
